src/compiler/crystal/semantic/main_visitor.cr

```diff
@@ -728,12 +728,24 @@ module Crystal
       end
     end
 
-    def end_visit(node : Expressions)
+    def visit(node : Expressions)
+      expressions = node.expressions
```

Comment:
Maybe there's a simpler way to write this? Iterate all expressions, including their index. Ignore type filters unless the index matches expressions size - es 1. I wonder if that within thing is slower than a regular each, that's all.